cpp/oneapi/dal/data/array.hpp

```diff
@@ -117,70 +160,132 @@ class array {
     void reset() {
         data_owned_ptr_.reset();
         data_ = std::variant<T*, const T*>();
-        size_ = 0;
+        count_ = 0;
         capacity_ = 0;
     }
 
-    void reset(std::int64_t size) {
-        data_owned_ptr_.reset(new T[size], default_delete{});
-        data_ = data_owned_ptr_.get();
-        size_ = size;
-        capacity_ = size;
+    void reset(std::int64_t count) {
+        reset_impl(detail::host_policy{}, count, detail::host_only_alloc{});
     }
 
     template <typename Deleter>
-    void reset(T* data, std::int64_t size, Deleter&& deleter) {
+    void reset(T* data, std::int64_t count, Deleter&& deleter) {
         // TODO: check input parameters
         data_owned_ptr_.reset(data, std::forward<Deleter>(deleter));
         data_ = data_owned_ptr_.get();
-        size_ = size;
-        capacity_ = size;
+        count_ = count;
+        capacity_ = count;
     }
 
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+    void reset(sycl::queue& queue,
+               std::int64_t count,
+               sycl::usm::alloc kind = sycl::usm::alloc::shared) {
+        reset_impl(queue, count, kind);
+    }
+
+    void reset(sycl::queue& queue,
+               T* data, std::int64_t count,
+               sycl::vector_class<sycl::event> dependencies = {}) {
+        reset(data, count, detail::default_delete<T, decltype(queue)>{ queue });
+        for (auto& event : dependencies) {
+            event.wait();
+        }
+    }
+#endif
+
     template <typename U = T*>
-    void reset_not_owning(U data = nullptr, std::int64_t size = 0) {
+    void reset_not_owning(U data = nullptr, std::int64_t count = 0) {
         data_ = data;
-        size_ = size;
+        count_ = count;
+    }
+
+    void resize(std::int64_t count) {
+        resize_impl(detail::host_policy{}, count, detail::host_only_alloc{});
+    }
+
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+    void resize(sycl::queue& queue,
+                std::int64_t count,
+                sycl::usm::alloc kind = sycl::usm::alloc::shared) {
+        resize_impl(queue, count, kind);
     }
+#endif
 
-    void resize(std::int64_t size) {
+    const T& operator [](std::int64_t index) const {
+        return get_data()[index];
+    }
+
+    T& operator [](std::int64_t index) {
+        return get_mutable_data()[index];
+    }
+
+private:
+    template <typename Policy, typename AllocKind>
+    static array<T> zeros_impl(Policy&& policy, std::int64_t count, AllocKind&& kind) {
+        auto* data = detail::malloc<T>(policy, count, kind);
+        detail::memset(policy, data, 0, sizeof(T)*count);
+        return array<T> { data, count, detail::default_delete<T, Policy>{ policy } };
+    }
+
+    template <typename K, typename Policy, typename AllocKind>
+    static array<T> full_impl(Policy&& policy,
```

Comment:
Why name is `full`, not `fill`? It confuses me a little :)

Reply:
It's from @rlnx request to make these functions like in numpy: https://numpy.org/doc/1.18/reference/generated/numpy.full.html
But numpy has ndarray.fill() function too. I can rename it if @rlnx don't mind.

Reply:
But I think that `full()` is more appropriate here since in numpy it constructs new ndarray, but `fill()` only fills existing one

Reply:
Ok, agree with you